sequencer/src/state_cert.rs

```diff
@@ -0,0 +1,119 @@
+//! State certificate validation and error handling
+
+use std::collections::HashMap;
+
+use alloy::primitives::{FixedBytes, U256};
+use anyhow::bail;
+use espresso_types::SeqTypes;
+use hotshot_query_service::availability::Error;
+use hotshot_task_impls::helpers::derive_signed_state_digest;
+use hotshot_types::{
+    light_client::StateVerKey,
+    simple_certificate::LightClientStateUpdateCertificateV2,
+    stake_table::HSStakeTable,
+    traits::signature_key::{LCV2StateSignatureKey, LCV3StateSignatureKey, StakeTableEntryType},
+};
+use tide_disco::StatusCode;
+
+/// Error type for state certificate fetching
+#[derive(Debug, thiserror::Error)]
+pub enum StateCertFetchError {
+    #[error("Failed to fetch state certificate: {0}")]
+    FetchError(#[source] anyhow::Error),
+
+    #[error("State certificate validation failed: {0}")]
+    ValidationError(#[source] anyhow::Error),
+
+    #[error("State certificate error: {0}")]
+    Other(#[source] anyhow::Error),
+}
+
+impl From<StateCertFetchError> for hotshot_query_service::availability::Error {
+    fn from(err: StateCertFetchError) -> Self {
+        match err {
+            StateCertFetchError::FetchError(e) => Error::Custom {
+                message: format!("Failed to fetch state cert from peers: {e}"),
+                status: StatusCode::NOT_FOUND,
+            },
+            StateCertFetchError::ValidationError(e) => Error::Custom {
+                message: format!("State certificate validation failed: {e}"),
+                status: StatusCode::INTERNAL_SERVER_ERROR,
+            },
+            StateCertFetchError::Other(e) => Error::Custom {
+                message: format!("Failed to process state cert: {e}"),
+                status: StatusCode::INTERNAL_SERVER_ERROR,
+            },
+        }
+    }
+}
+
+/// Validates a state certificate by verifying signatures and checking threshold
+pub fn validate_state_cert(
+    cert: &LightClientStateUpdateCertificateV2<SeqTypes>,
+    stake_table: &HSStakeTable<SeqTypes>,
+) -> anyhow::Result<()> {
+    let signed_state_digest = derive_signed_state_digest(
+        &cert.light_client_state,
+        &cert.next_stake_table_state,
+        &cert.auth_root,
+    );
+
+    // If auth_root is the default value (all zeros), we're on consensus version V3, so verify LCV2 signatures
+    // Otherwise, verify LCV3 signatures
+    let use_lcv2 = cert.auth_root == FixedBytes::<32>::default();
```

Comment:
We had a little mis communication here, for consensus V4 we should verify both LCV3 signature and LCV2 signature.